### PR TITLE
Monitor RAM and flash disk usage in Mikrotik routers

### PR DESCRIPTION
**Pull request by alejandroscf**
Two scripts to monitor RAM and flash disk usage in Mikrotik routers via SNMP protocol

### CODE REVIEW

File: plugins/router/snmp__mikrotik_flash [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+
+if ($ARGV[0] and $ARGV[0] eq "config")
+{
+    print "host_name $host\n";
+    if (!defined ($response = $session->get_request($sysRAMTotal)))
+    {
+        die "Croaking: " . $session->error();
+    }
+#    print "graph_args --base 1000 -r --lower-limit 0 --upper-limit 100\n";
```
Comment (sumpfralle) [inline plugins/router/snmp__mikrotik_flash]: Please remove commented code.

File: plugins/router/snmp__mikrotik_flash [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    print "sysRAMTotal.draw AREA\n";
+    print "sysRAMUsage.label Used Memory\n";
+    print "sysRAMUsage.draw AREA\n";
+    exit 0;
+}
+
+
+if (defined ($response = $session->get_request(-varbindlist => [$sysRAMUsage, $sysRAMTotal])))
+{
+        print "sysRAMUsage.value ", $response->{$sysRAMUsage}*1024, "\n";
```
Comment (sumpfralle) [inline plugins/router/snmp__mikrotik_flash]: Is this a conversion from bytes to kilobytes? Usually SI base units are preferable in order to let munin/rrd print proper units for higher values.

Another detail: is `RAM` the proper name here? Shouldn't it be about the flash storage?

File: plugins/router/snmp__mikrotik_flash [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        print "sysRAMTotal.value ", $response->{$sysRAMTotal}*1024, "\n";
+}
+else
+{
+        print "sysRAMUsage.value U\n";
```
Comment (sumpfralle) [inline plugins/router/snmp__mikrotik_flash]: I guess, you should also emit `sysRAMTotal` as `U`?

File: plugins/router/snmp__mikrotik_flash [hunk 4]
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/perl -w
+#
```
Comment (sumpfralle) [inline plugins/router/snmp__mikrotik_flash]: Please use a perldoc header here - this allows automatic man page generation and better visualization in the plugin gallery.

File: plugins/router/snmp__mikrotik_ram [hunk 5]
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/perl -w
+#
```
Comment (sumpfralle) [inline plugins/router/snmp__mikrotik_ram]: Perldoc header, please.

File: plugins/router/snmp__mikrotik_ram [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+
+if ($ARGV[0] and $ARGV[0] eq "config")
+{
+    print "host_name $host\n";
+    if (!defined ($response = $session->get_request($sysRAMTotal)))
+    {
+        die "Croaking: " . $session->error();
+    }
+#    print "graph_args --base 1000 -r --lower-limit 0 --upper-limit 100\n";
```
Comment (sumpfralle) [inline plugins/router/snmp__mikrotik_ram]: no commented code, please.

File: plugins/router/snmp__mikrotik_ram [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    print "sysRAMTotal.draw AREA\n";
+    print "sysRAMUsage.label Used Memory\n";
+    print "sysRAMUsage.draw AREA\n";
+    exit 0;
+}
+
+
+if (defined ($response = $session->get_request(-varbindlist => [$sysRAMUsage, $sysRAMTotal])))
+{
+        print "sysRAMUsage.value ", $response->{$sysRAMUsage}*1024, "\n";
```
Comment (sumpfralle) [inline plugins/router/snmp__mikrotik_ram]: Same as above: is the conversion necessary?

File: plugins/router/snmp__mikrotik_ram [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        print "sysRAMTotal.value ", $response->{$sysRAMTotal}*1024, "\n";
+}
+else
+{
+        print "sysRAMUsage.value U\n";
```
Comment (sumpfralle) [inline plugins/router/snmp__mikrotik_ram]: Same here: missing `total`?

File: plugins/router/snmp__mikrotik_flash (hunk 2 above)
Comment (alejandroscf) [reply plugins/router/snmp__mikrotik_flash]: No, the conversion is from kilobytes to bytes. SNMP is returning it in kilobytes so I convert it to bytes in order to had it in SI baseunits. 

Yes, that should be "FLASH" or something like that.

File: plugins/router/snmp__mikrotik_flash (hunk 1 above)
Comment (alejandroscf) [reply plugins/router/snmp__mikrotik_flash]: Sure

File: plugins/router/snmp__mikrotik_flash (hunk 3 above)
Comment (alejandroscf) [reply plugins/router/snmp__mikrotik_flash]: Yes, it should.

File: plugins/router/snmp__mikrotik_flash (hunk 4 above)
Comment (alejandroscf) [reply plugins/router/snmp__mikrotik_flash]: I'm not sure how to use perldoc. It's like the headers in other plugins?

**Comment by alejandroscf** [comment]
Thank for you comments. I'll try to fix that.
I made it in two different files so you can choose what to monitor in each situation. For example I use this for work and only monitor flash usage in some routers, but I want to keep an eye in RAM usage in all of them. Also the snmp_zyxel_zywall__ plugins also had it in different files.

File: plugins/router/snmp__mikrotik_flash (hunk 4 above)
Comment (sumpfralle) [reply plugins/router/snmp__mikrotik_flash]: Yes. Just look for `=cut` in other plugins.

**Comment by sumpfralle** [comment]
> For example I use this for work and only monitor flash usage in some routers, but I want to keep an eye in RAM usage in all of them.

Specifc graphs can also be excluded by configuration. Maybe this would be a better solution?

> Also the snmp_zyxel_zywall__ plugins also had it in different files.

Yes, but I would like to prevent bad examples (code copies) from spreading :)

Regarding: the SNMP OIDs: do you think, they are applicable for other SNMP devices? Thus maybe the plugin name could be more generic?

**Comment by sumpfralle** [comment]
Ping?

File: plugins/router/snmp__mikrotik_flash (hunk 4 above)
Comment (alejandroscf) [reply plugins/router/snmp__mikrotik_flash]: I've added the header, I hope it's ok.

**Comment by alejandroscf** [comment]
I fixed the misspelling and the Perl Doc header. I'm sorry but I have no time to solve the multigraph thing, maybe we could merge now?

**Comment by sumpfralle** [comment]
Thank you!

I really do not like to introduce code duplication into the repository. Thus I took the liberty to migrate the plugins to the multigraph approach myself: #1047.

Please report back whether the unified plugin works for you.

**Comment by alejandroscf** [comment]
Thank you very much! 

I've just tested it and I left you a suggestion in the other PR so I'm closing this one.

**Pull request closed without merging**